chalk-solve/src/clauses/builtin_traits/fn_.rs

```diff
@@ -0,0 +1,93 @@
+use crate::clauses::ClauseBuilder;
+use crate::infer::instantiate::IntoBindersAndValue;
+use crate::rust_ir::WellKnownTrait;
+use crate::{Interner, RustIrDatabase, TraitRef};
+use chalk_ir::{
+    AliasTy, ApplicationTy, Binders, Normalize, ProjectionTy, Substitution, TraitId, Ty, TyData,
+    TypeName, VariableKinds,
+};
+
+/// Handles clauses for FnOnce/FnMut/Fn.
+/// If `assoc_output` is `true`, we push a clause of the form
+/// `Normalize(<fn(A) -> B as FnOnce<(A,)>>::Output -> B) :- Implemented(fn(A) -> B as FnOnce<(A,)>`
+///
+/// If `assoc_output` is `false`, we push a clause of the form
+/// `Implemented(fn(A) -> B as FnOnce<(A,)>)`
+pub fn add_fn_trait_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_id: TraitId<I>,
+    ty: &TyData<I>,
+    assoc_output: bool,
```

Comment:
I think instead of a `bool` here, we can just pass the `WellKnownTrait` and if `FnOnce`, *also* add the `Normalize` clause. `CouldMatch` can just decide which one to use later.

This will also be needed later on for the closure kind.